charmhelpers/contrib/storage/linux/ceph.py

```diff
@@ -856,19 +856,35 @@ def _keyring_path(service):
     return KEYRING.format(service)
 
 
-def create_keyring(service, key):
-    """Create a new Ceph keyring containing key."""
+def add_key(service, key):
+    """
+    Add a key to a keyring.
+
+    Creates the keyring if it doesn't already exist.
+
+    Logs and returns if the key is already in the keyring.
+    """
     keyring = _keyring_path(service)
     if os.path.exists(keyring):
-        log('Ceph keyring exists at %s.' % keyring, level=WARNING)
-        return
+        with open(keyring) as ring:
```

Comment:
My py3 concern still stands. Will this be bytes and require decoding?

We should be explicit if this is 'r' or 'rb' and decode if necessary.

Reply:
I spun up an openstack base bundle in serverstack. I was able to read all the .keyring files that I found, in python3, with an open("filename", "r"), and check to see if a given string existed in the file. I don't think that we have to worry about unicode encoding. :-)